quantum/process_keycode/process_autocorrect.c

```diff
@@ -0,0 +1,247 @@
+    uint8_t        mods                                = get_mods();
+#ifndef NO_ACTION_ONESHOT
+    mods |= get_oneshot_mods();
+#endif
+
+    if ((keycode >= AUTOCORRECT_ON && keycode <= AUTOCORRECT_TOGGLE) && record->event.pressed) {
+        if (keycode == AUTOCORRECT_ON) {
+            keymap_config.autocorrect_enable = true;
+        } else if (keycode == AUTOCORRECT_OFF) {
+            keymap_config.autocorrect_enable = false;
+        } else if (keycode == AUTOCORRECT_TOGGLE) {
+            keymap_config.autocorrect_enable = !keymap_config.autocorrect_enable;
+        } else {
+            return true;
+        }
+
+        typo_buffer_size = 0;
+        eeconfig_update_keymap(keymap_config.raw);
+        return false;
+    }
+
+    if (!keymap_config.autocorrect_enable) {
+        typo_buffer_size = 0;
+        return true;
+    }
+
+    if (!record->event.pressed) {
+        return true;
+    }
+
+    // autocorrect keycode verification and extraction
+    if (!process_autocorrect_user(&keycode, record, &typo_buffer_size, &mods)) {
+        return true;
+    }
+
+    // keycode buffer check
+    switch (keycode) {
+        case KC_A ... KC_Z:
+            // process normally
+            break;
+        case KC_1 ... KC_0:
+        case KC_TAB ... KC_SEMICOLON:
+        case KC_GRAVE ... KC_SLASH:
+            // Set a word boundary if space, period, digit, etc. is pressed.
+            keycode = KC_SPC;
+            break;
+        case KC_ENTER:
+            // Behave more conservatively for the enter key. Reset, so that enter
+            // can't be used on a word ending.
+            typo_buffer_size = 0;
+            keycode          = KC_SPC;
+            break;
+        case KC_BSPC:
+            // Remove last character from the buffer.
+            if (typo_buffer_size > 0) {
+                --typo_buffer_size;
+            }
+            return true;
+        case KC_QUOTE:
+            // Treat " (shifted ') as a word boundary.
+            if ((mods & MOD_MASK_SHIFT) != 0) {
+                keycode = KC_SPC;
+            }
+            break;
+        default:
+            // Clear state if some other non-alpha key is pressed.
+            typo_buffer_size = 0;
+            return true;
+    }
+
+    // Rotate oldest character if buffer is full.
+    if (typo_buffer_size >= AUTOCORRECT_MAX_LENGTH) {
+        memmove(typo_buffer, typo_buffer + 1, AUTOCORRECT_MAX_LENGTH - 1);
+        typo_buffer_size = AUTOCORRECT_MAX_LENGTH - 1;
+    }
+
+    // Append `keycode` to buffer.
+    typo_buffer[typo_buffer_size++] = keycode;
+    // Return if buffer is smaller than the shortest word.
+    if (typo_buffer_size < AUTOCORRECT_MIN_LENGTH) {
+        return true;
+    }
+
+    // Check for typo in buffer using a trie stored in `autocorrect_data`.
+    uint16_t state = 0;
+    uint8_t  code  = pgm_read_byte(autocorrect_data + state);
+    for (uint8_t i = typo_buffer_size - 1; i >= 0; --i) {
```

Comment:
@filterpaper You mean like this? 
```suggestion
    for (uint8_t i = typo_buffer_size; i-- >= 0;) {
```

Reply:
I think using `int` in QMK core will less obfuscated in this format: 
```c
for (int i = typo_buffer_size - 1; i >= 0; --i) {
```
Better clarity at the expense of a few additional bytes.

Reply:
@filterpaper great catch! Yes, `uint8_t` is a tricky data type to make a loop that counts down. Signed intere. I used `int` in my code and overlooked that it was different here.

I suggest to use `int8_t` for the loop counter and add a check in the Python code that typos aren't over 127 chars.

```py
if len(typo) > 127:
    print(f'Error:{line_number}: typo is exceeds 127 chars: "{typo}"')
    sys.exit(1)
```

Reply:
I've been bitten before by looping unsigned integer and should have paid better attention. My apologies for the false alarm with `state`. Agree that `uint8_t` is a better choice although it didn't result with a smaller compiled size.